tests/neg/local-implicit-function.scala

```diff
@@ -0,0 +1,45 @@
+object Test {
+  type using[A,E] = implicit local E => A
+
+  case class Incrementor(i1: Int) {
+    def incr(i2: Int) = i1 + i2
+  }
+
+  def incr(i: Int): Int using Incrementor = implicitly[Incrementor].incr(i)
+
+  implicit local val inc: Incrementor = Incrementor(2)
+
+  def calculation(): Int using Incrementor = incr(40) // OK
+
+  // Try to store
+  var incMut: Incrementor = Incrementor(42)
+  def returnId(implicit ina: Incrementor): Incrementor = ina
+  def escape1(): Unit using Incrementor = implicitly[Incrementor] // TODO: should error
+  def escape2(): Unit using Incrementor = incMut = returnId // error
+
+
+  // A basic capability
+
+  case class ExcCap(fn: Exception => Nothing)
+  def trying[A](fn: implicit local ExcCap => A): Option[A] =
+    Some(fn(ExcCap(x => return None)))
+  def throwing(e: Exception)(implicit local excCap: ExcCap) =
+    excCap.fn(e)
+  def runUnsafe[A](fn: implicit () => A) = fn()
+  def runSafe[A](local fn: implicit () => A) = fn()
+
+  // legal
+  trying {
+    runSafe {
+      throwing(new Exception);
+      print("Foo")
+    }
+  }
+
+  // not legal due to usage of unsafe
+  trying {
+    runUnsafe {
+      throwing(new Exception) // error
+    }
+  }
+}
```

Comment:
This example is based on one from *the gentrification paper*.
With the help of implicit (local) function types this is now even niftier.

Old:
```scala
trying { throwing =>
  runUnsafe {
    () => throwing(new Exception) // error
   }
}
```

New:
```scala
trying {
  runUnsafe {
    throwing(new Exception) // error
   }
}
```
:tada:

Reply:
That's what I am talking about ;)

Reply:
Currently bootstrapping the Dotty compiler fails, since `local` is used as a name, for instance in `Desugar.scala:90:16`

```scala
var local = defctx.denotNamed(tp.name ++ suffix).suchThat(_.isParamOrAccessor).symbol
```

Reply:
Good point!

Should we try to demote `local` and `safe` from keywords to identifiers for better compatibility?

Reply:
Not sure what's the best strategy here, but I would be fine with keeping them as keywords and renaming usages to `l` or something else.

Reply:
(As long as the renaming happens in one commit ;) )

Reply:
Another side-note: It looks like `local` makes currying a bit akward:

```scala
f: implicit local A => implicit local B => C
//   =?=
g: implicit local (A, B) => C
```
in function `f` we will close over the local of type `A`, which will be rejected.